File: app/src/main/java/com/yagi/android/numbercatch/MainActivity.java
package com.yagi.android.numbercatch;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextPaint;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import java.util.List;
import java.util.Random;

public class MainActivity extends AppCompatActivity implements SensorEventListener {

    GameView gameView;
    SensorManager sensorManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        SurfaceView surfaceView = (SurfaceView) findViewById(R.id.SurfaceViewMain);
        gameView = new GameView(this, surfaceView);
        setContentView(gameView);
        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.menu_restart) {
            gameView.thread.interrupt();
            AlertDialog.Builder alertDlg = new AlertDialog.Builder(this);
            alertDlg.setTitle("終了しますか？");
            alertDlg.setMessage("プレイ中の記録は戻らなくなります");
            alertDlg.setPositiveButton(
                    "はい",
                    new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            // OK ボタンクリック処理
                            Intent i = new Intent(MainActivity.this, MainActivity.class);
                            startActivity(i);
                            finish();
                        }
                    });
            alertDlg.setNegativeButton(
                    "いいえ",
                    new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            // Cancel ボタンクリック処理

                        }
                    });

            // 表示
            alertDlg.create().show();


        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        List<Sensor> sensors = sensorManager.getSensorList(Sensor.TYPE_ACCELEROMETER);
        if (!sensors.isEmpty()) {
            sensorManager.registerListener(this, sensors.get(0), SensorManager.SENSOR_DELAY_FASTEST);
        }
    }

    @Override
    protected void onPause() {
        sensorManager.unregisterListener(this);
        super.onPause();
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            Log.d("SensorValues", "\nX軸:" + event.values[0] + "\nY軸" + event.values[1] + "\nZ軸" + event.values[2]);
            if (gameView.player != null) {
                gameView.player.move(-event.values[0]);
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }


    public class GameView extends SurfaceView implements SurfaceHolder.Callback, Runnable {

        int screenWidth, screenHeight;
        int score = 0;
        int life = 400;
        int[] fallingNumbers = new int[5];
        int countUp;

        static final long FPS = 30;
        static final long FRAME_TIME = 1000 / FPS;

        SurfaceHolder surfaceHolder;
        Thread thread;
        Number[] numbers;
        Bitmap[] numberImages;
        Matrix matrix = new Matrix();
        Matrix matrix2 = new Matrix();
        Player player;
        MainActivity mMainActivity;


        public GameView(Context context, SurfaceView sv) {
            super(context);
            numberImages = new Bitmap[5];
            this.mMainActivity = (MainActivity) context;
            getHolder().addCallback(this);
            float presentImage_ratio_w = 0.05f;
            float presentImage_ratio_h = 0.05f;
            float playerImage_ratio_w = 0.08f;
            float playerImage_ratio_h = 0.08f;
            matrix.postScale(presentImage_ratio_w, presentImage_ratio_h);
            matrix2.postScale(playerImage_ratio_w, playerImage_ratio_h);
            Random r = new Random();
            for (int in = 0; in < fallingNumbers.length; in++) {
                fallingNumbers[in] = r.nextInt(30) + 1;
            }
        }

        @Override
        public void surfaceCreated(SurfaceHolder holder) {
            surfaceHolder = holder;
            thread = new Thread(this);
            thread.start();
        }

        @Override
        public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            screenWidth = width;
            screenHeight = height;
        }

        @Override
        public void surfaceDestroyed(SurfaceHolder holder) {
            thread = null;
        }


        @Override
        public void run() {
            int counter = 0;
            player = new Player();
            numbers = new Number[5];
            numbers[0] = new Number(0);
            numbers[1] = new Number(1);
            numbers[2] = new Number(2);
            numbers[3] = new Number(3);
            numbers[4] = new Number(4);
            boolean[] pre = new boolean[5];
            pre[0] = true;
            for (int i = 1; i < fallingNumbers.length; i++) {
                pre[i] = false;
            }

            while (thread != null) {
                final Canvas canvas = surfaceHolder.lockCanvas();
                canvas.drawColor(Color.WHITE);
                final Paint catchPaint = new Paint();
                Paint lifePaint = new Paint();
                catchPaint.setTextSize(120);
                lifePaint.setTextSize(100);
                canvas.drawText(String.valueOf(score), player.x, player.y, catchPaint);
                Paint[] textPaint = new Paint[5];
                for (int i = 0; i < textPaint.length; i++) {
                    textPaint[i] = new TextPaint(i);
                    textPaint[i].setTextSize(100);
                }
                textPaint[0].setColor(Color.rgb(255, 70, 0));
                textPaint[1].setColor(Color.BLUE);
                textPaint[2].setColor(Color.rgb(34, 139, 34));
                textPaint[3].setColor(Color.MAGENTA);
                textPaint[4].setColor(Color.GRAY);
                canvas.drawText(String.valueOf(fallingNumbers[0]), numbers[0].x, numbers[0].y, textPaint[0]);
                for (int i = 0; i < fallingNumbers.length; i++) {
                    if (counter > i * 5 - 1) {
                        canvas.drawText(String.valueOf(fallingNumbers[i]), numbers[i].x, numbers[i].y, textPaint[i]);
                        pre[i] = true;
                    }
                }

                for (int in = 0; in < fallingNumbers.length; in++) {
                    if (pre[in]) {
                        if (player.isEnter(numbers[in])) {
                            countUp = score;
                            score += fallingNumbers[in];
                            numbers[in].reset(in);
                            counter++;
                        } else if (numbers[in].y > screenHeight) {
                            life -= fallingNumbers[in];
                            numbers[in].reset(in);
                        } else {
                            numbers[in].update();
                        }
                    }

                }

                canvas.drawText(String.valueOf(score), player.x, player.y, catchPaint);
                for (Paint paint : textPaint) {
                    paint.setFakeBoldText(true);
                }


                if (life <= 0) {
                    life = 0;
                    lifePaint.setColor(Color.RED);
                    canvas.drawText("LIFE:" + 0, 50, 150, lifePaint);
                    canvas.drawText("Game Over", screenWidth / 3, screenHeight / 2, lifePaint);
                    surfaceHolder.unlockCanvasAndPost(canvas);

                    Intent i = new Intent(mMainActivity, FinishActivity.class);
                    Log.d("intent", String.valueOf(score));
                    i.putExtra("score", score);
                    startActivity(i);

                    break;

                } else if (life <= 100) {
                    lifePaint.setColor(Color.RED);
                    canvas.drawText("LIFE:" + life, 50, 150, lifePaint);
                } else {
                    canvas.drawText("LIFE:" + life, 50, 150, lifePaint);
                }
                surfaceHolder.unlockCanvasAndPost(canvas);

                try {
                    Thread.sleep(FRAME_TIME);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }

        class Number {
            private static final int WIDTH = 100;
            private static final int HEIGHT = 100;

            float x, y;


            private void update() {
                y += 30.0f;
            }

            private Number(int in) {
                Random random = new Random();
                x = random.nextInt(screenWidth - WIDTH);
                y = 0;
                Random r = new Random();
                fallingNumbers[in] = r.nextInt(30) + 1;
            }


            private void reset(int in) {
                Random random = new Random();
                x = random.nextInt(screenWidth - WIDTH);
                y = 0;
                Random r = new Random();
                fallingNumbers[in] = r.nextInt(30) + 1;
            }
        }

        class Player {
            final int WIDTH = 200;
            final int HEIGHT = 200;

            float x, y;

            private Player() {
                x = 0;
                y = screenHeight - HEIGHT;
            }

            private void move(float diffX) {
                this.x += diffX;
                this.x = Math.max(0, x);
                this.x = Math.min(screenWidth - WIDTH, x);
            }

            private boolean isEnter(Number number) {
                return number.x + Number.WIDTH > x && number.x < x + WIDTH && number.y + Number.HEIGHT > y && number.y < y + HEIGHT;
            }
        }
    }

    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {

            // 確認ダイアログの生成
            AlertDialog.Builder alertDlg = new AlertDialog.Builder(this);
            alertDlg.setTitle("終了しますか？");
            alertDlg.setMessage("プレイ中の記録は戻らなくなります");
            alertDlg.setPositiveButton(
                    "はい",
                    new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            // OK ボタンクリック処理
                            finish();
                        }
                    });
            alertDlg.setNegativeButton(
                    "いいえ",
                    new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            // Cancel ボタンクリック処理

                        }
                    });

            // 表示
            alertDlg.create().show();
            return true;


        }

        return super.onKeyDown(keyCode, event);
    }


}

File: app/src/main/java/com/yagi/android/numbercatch/MainActivity.kt
package com.yagi.android.numbercatch

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.text.TextPaint
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import java.util.*
import android.graphics.RectF
import kotlin.math.max


class MainActivity : AppCompatActivity(), SensorEventListener {
    var gameView: GameView? = null
    var sensorManager: SensorManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        findViewById<SurfaceView>(R.id.SurfaceViewMain)
        gameView = GameView(this)
        setContentView(gameView)
        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.menu_restart) {
            gameView!!.thread!!.interrupt()
            val alertDlg = AlertDialog.Builder(this)
            alertDlg.setTitle("ゲームをリスタートしますか？")
            alertDlg.setMessage("プレイ中の記録は戻らなくなります")
            alertDlg.setPositiveButton(
                "はい"
            ) { dialog, which -> // OK ボタンクリック処理
                val i = Intent(this@MainActivity, MainActivity::class.java)
                startActivity(i)
                finish()
            }
            alertDlg.setNegativeButton(
                "いいえ"
            ) { dialog, which ->
                // Cancel ボタンクリック処理
            }

            // 表示
            alertDlg.create().show()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        val sensors = sensorManager!!.getSensorList(Sensor.TYPE_ACCELEROMETER)
        if (!sensors.isEmpty()) {
            sensorManager!!.registerListener(this, sensors[0], SensorManager.SENSOR_DELAY_FASTEST)
        }
    }

    override fun onPause() {
        sensorManager!!.unregisterListener(this)
        super.onPause()
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type == Sensor.TYPE_ACCELEROMETER) {
            if (gameView!!.player != null) {
                gameView!!.player!!.move(-event.values[0])
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    inner class GameView(context: Context?) : SurfaceView(context),
        SurfaceHolder.Callback, Runnable {
        var screenWidth = 0
        var screenHeight = 0
        var score = 0
        var life = 400
        var fallingNumbers = IntArray(5)
        var countUp = 0
        var surfaceHolder: SurfaceHolder? = null
        var thread: Thread? = null
        lateinit var numbers: Array<Number?>
        var numberImages: Array<Bitmap?>
        var matrix1 = Matrix()
        var matrix2 = Matrix()
        var player: Player? = null
        var mMainActivity: MainActivity?
        override fun surfaceCreated(holder: SurfaceHolder) {
            surfaceHolder = holder
            thread = Thread(this)
            thread!!.start()
        }

        override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
            screenWidth = width
            screenHeight = height
        }

        override fun surfaceDestroyed(holder: SurfaceHolder) {
            thread = null
        }

        override fun run() {
            var counter = 0
            player = Player()
            numbers = arrayOfNulls(5)
            numbers[0] = Number(0)
            numbers[1] = Number(1)
            numbers[2] = Number(2)
            numbers[3] = Number(3)
            numbers[4] = Number(4)
            val pre = BooleanArray(5)
            pre[0] = true
            for (i in 1 until fallingNumbers.size) {
                pre[i] = false
            }
            while (thread != null) {
                val canvas = surfaceHolder!!.lockCanvas()
                canvas.drawColor(Color.WHITE)
                val catchPaint = Paint()
                val lifePaint = Paint()
                val lifeBackPaint = Paint()
                catchPaint.textSize = 120f
                lifePaint.textSize = 100f
                canvas.drawText(score.toString(), player!!.x, player!!.y, catchPaint)
                val textPaint = arrayOfNulls<Paint>(5)
                for (i in textPaint.indices) {
                    textPaint[i] = TextPaint(i)
                    textPaint[i]?.setTextSize(100f)
                }
                textPaint[0]!!.color = Color.rgb(255, 70, 0)
                textPaint[1]!!.color = Color.BLUE
                textPaint[2]!!.color = Color.rgb(34, 139, 34)
                textPaint[3]!!.color = Color.MAGENTA
                textPaint[4]!!.color = Color.GRAY
                canvas.drawText(
                    fallingNumbers[0].toString(),
                    numbers[0]!!.x,
                    numbers[0]!!.y,
                    textPaint[0]!!
                )
                for (i in fallingNumbers.indices) {
                    if (counter > i * 5 - 1) {
                        canvas.drawText(
                            fallingNumbers[i].toString(),
                            numbers[i]!!.x,
                            numbers[i]!!.y,
                            textPaint[i]!!
                        )
                        pre[i] = true
                    }
                }
                for (`in` in fallingNumbers.indices) {
                    if (pre[`in`]) {
                        if (player!!.isEnter(numbers[`in`])) {
                            countUp = score
                            score += fallingNumbers[`in`]
                            numbers[`in`]!!.reset(`in`)
                            counter++
                        } else if (numbers[`in`]!!.y > screenHeight) {
                            life = max(life - fallingNumbers[`in`], 0)
                            numbers[`in`]!!.reset(`in`)
                        } else {
                            numbers[`in`]!!.update()
                        }
                    }
                }
                canvas.drawText(score.toString(), player!!.x, player!!.y, catchPaint)
                for (paint in textPaint) {
                    paint!!.isFakeBoldText = true
                }
                val rectF = RectF(0F, 0F, screenWidth.toFloat(), 200F)
                lifeBackPaint.setColor(Color.rgb(240, 240, 240))
                canvas.drawRect(rectF, lifeBackPaint)
                if (life <= 100){
                    lifePaint.color = Color.RED
                }
                canvas.drawText("LIFE:$life", 50f, 135f, lifePaint)
                if (life == 0) {
                    canvas.drawText(
                        "Game Over",
                        (screenWidth / 3).toFloat(),
                        (screenHeight / 2).toFloat(),
                        lifePaint
                    )
                    surfaceHolder!!.unlockCanvasAndPost(canvas)
                    val i = Intent(mMainActivity, FinishActivity::class.java)
                    i.putExtra("score", score)
                    startActivity(i)
                    break
                }
                surfaceHolder!!.unlockCanvasAndPost(canvas)
                try {
                    Thread.sleep(FRAME_TIME)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }

        inner class Number(`in`: Int) {
            var x: Float
            var y: Float
            fun update() {
                y += 30.0f
            }

            val WIDTH = 100
            val HEIGHT = 100


            fun reset(`in`: Int) {
                val random = Random()
                x = random.nextInt(screenWidth - WIDTH).toFloat()
                y = 0f
                val r = Random()
                fallingNumbers[`in`] = r.nextInt(30) + 1
            }

            init {
                val random = Random()
                x = random.nextInt(screenWidth - WIDTH).toFloat()
                y = 0f
                val r = Random()
                fallingNumbers[`in`] = r.nextInt(30) + 1
            }
        }

        inner class Player {
            val WIDTH = 200
            val HEIGHT = 200
            var x = 0f
            var y: Float
            fun move(diffX: Float) {
                x += diffX
                x = Math.max(0f, x)
                x = Math.min((screenWidth - WIDTH).toFloat(), x)
            }

            fun isEnter(number: Number?): Boolean {
                return number!!.x + number.WIDTH > x && number.x < x + WIDTH && number.y + number.HEIGHT > y && number.y < y + HEIGHT
            }

            init {
                y = (screenHeight - HEIGHT).toFloat()
            }
        }


        private val FPS: Long = 30
        val FRAME_TIME = 1000 / FPS


        init {
            numberImages = arrayOfNulls(5)
            mMainActivity = context as MainActivity?
            holder.addCallback(this)
            val presentImage_ratio_w = 0.05f
            val presentImage_ratio_h = 0.05f
            val playerImage_ratio_w = 0.08f
            val playerImage_ratio_h = 0.08f
            matrix1.postScale(presentImage_ratio_w, presentImage_ratio_h)
            matrix2.postScale(playerImage_ratio_w, playerImage_ratio_h)
            val r = Random()
            for (`in` in fallingNumbers.indices) {
                fallingNumbers[`in`] = r.nextInt(30) + 1
            }
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {

            // 確認ダイアログの生成
            val alertDlg = AlertDialog.Builder(this)
            alertDlg.setTitle("終了しますか？")
            alertDlg.setMessage("プレイ中の記録は戻らなくなります")
            alertDlg.setPositiveButton(
                "はい"
            ) { dialog, which -> // OK ボタンクリック処理
                finish()
            }
            alertDlg.setNegativeButton(
                "いいえ"
            ) { dialog, which ->
                // Cancel ボタンクリック処理
            }

            // 表示
            alertDlg.create().show()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }
}